sycl/doc/extensions/proposed/sycl_ext_oneapi_graph_fusion.asciidoc

```diff
@@ -0,0 +1,644 @@
+optimizations such as dataflow internalization discussed below.
+
+This proposal is a continuation of many of the ideas of the initial
+https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_codeplay_kernel_fusion.asciidoc[experimental
+kernel fusion proposal] for SYCL. However, instead of defining its own
+SYCL-based API to record a sequence of kernels to fuse, this proposal builds on
+top of the graphs proposal to allow the fusion of graphs. This not only unifies
+the APIs, making sure users only need to familiarize themselves with a single
+API, but also provides additional advantages. 
+
+The graph proposal defines two APIs to create graphs: a proposal using a
+recording mechanism, similar to the initial kernel fusion proposal; and another
+one using explicit graph building. Thus, future users will be able to choose
+from two different mechanisms to construct the sequence of kernels to fuse. As
+there is an explicit step for finalization of graphs before being submitted for
+execution, fusion can happen in this step, which also eliminates many of the
+synchronization concerns that needed to be covered in the experimental kernel
+fusion proposal.
+
+The aim of this document is to propose a mechanism for users to request the
+fusion of two or more kernels in a SYCL graph into a single kernel **at
+runtime**. This requires the extension of the runtime with some sort of JIT
+compiler to allow for the fusion of kernel functions at runtime.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_GRAPH_FUSION` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to determine
+if the implementation supports this feature, or applications can test the
+macro's value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== API modifications
+
+==== Properties
+
+===== Graph Fusion Property
+
+The API for `command_graph<graph_state::modifiable>::finalize()` includes a
+`property_list` parameter. The following property, defined by this extension,
+can be added to the property list to indicate that the kernels in the
+command-graph should be fused. 
+
+```c++
+sycl::ext::oneapi::experimental::property::command_graph::perform_fusion 
+```
+
+The property is not prescriptive. Implementations are free to not perform fusion
+if it is not possible (see below section <<limitations, Limitations>>), fusion is not
+supported by the implementation, or the implementation decides not to perform
+fusion for other reasons. It is not an error if an implementation does not
+perform fusion even though the property is passed. 
```

Comment:
This is just my opinion, but I find it a bit weird that this interface requires developers to provide lots of decorations and properties about what they want to happen, but that there's still so much implementation-defined behavior.

I think it would be preferable to separate "make a best effort to perform fusion" from "perform fusion", where the former is allowed to fail spuriously and the latter throws an exception if it fails.  The former might already be covered by as-if optimization guarantees.

I say this because if somebody is trying to optimize a code and prepare a graph for fusion, they should know when the property doesn't do what they asked.

Reply:
> For the internalization properties (promote_local/promote_private), I think they should remain to not be prescriptive. 

I agree.  These should be hints that give the compiler information about which kernels could legally be fused.

> For the fusion itself, we could make this prescriptive. If an implementation reports support for fusion on a device through the aspect/information descriptor, we could mandate a synchronous error to be thrown if fusion is not happening.

This makes sense to me.

It seems to me like an implementation _could_ choose to try and fuse all of the kernels in a graph without any input from the user: as long as the implementation produces the right answer (and respects the properties attached to buffers, etc), the user can't really tell if fusion happened or not.  If I'm right about that, then the descriptive version of `perform_fusion` doesn't really add very much (except that our current implementation doesn't attempt to fuse if it's not present).

If we want to keep a descriptive version around, I'd recommend the naming `enable_fusion` (similar to `enable_profiling`).  It suggests that the implementation _may_ fuse, but isn't required to.  It might be a no-op, either because an implementation always fuses, or because it never fuses.

For a prescriptive version, I'd recommend something like `require_fusion`.  It could coexist with the `enable_` variant.

> Would you also want to define a minimum set of fusion scenarios that must be supported for an implementation to be allowed to report fusion support?

I don't think we have to go this far, at least right now.  I think `require_fusion` would have to mean something like "maximally fuse the linearization of this graph".  In the absence of enqueued barriers, that probably means making everything into one kernel?  With enqueued barriers, I guess you'd get one kernel between each pair of barriers.

Any implementation that can't fuse all the kernels together (for any reason) would throw an exception.  The user could then enqueue again without fusion, or investigate why fusion isn't happening, depending on their use-case.

Reply:
That sounds reasonable to me. 

The necessity of a descriptive variant hinges on the guarantees graphs want give with regard to outside visibility of individual kernels. If SYCL graphs want to guarantee that tools such as profilers (Intel VTune, Nvidia NSight, ...) can observe individual kernels, fusion would only be possible if explicitly allowed by the user through a descriptive property. If, on the other hand, SYCL graphs give no such guarantee, we could omit the descriptive variant and implementations are free to perform fusion without explicit instruction to do so. 

What type of guarantees should/do graphs give in that regard?

Independent of that, we would have a prescriptive version, throwing an error if fusion fails.

Reply:
That's a very good question.  I don't know the answer, so tagging some folks who might know and/or have opinions: @reble, @bashbaug, @gmlueck 

I can see arguments both ways... But I agree that the necessity of the descriptive variant hinges on the answer to this question.

Reply:
The examples you give for "outside observability" are all profilers.  It seems unlikely to me that a user would want to disable optimizations (like fusion) when running a profiler.  Wouldn't a user want to profile the optimized code?  Instead, maybe the graph fusion implementation should name the fused kernel in a way that makes it obvious which input kernels it corresponds to.

Are there any issues with "inside observability", though?  For example, what if the kernel has a type-name, and the application gets a `kernel_bundle` from that type-name?